### DISCUSSION TITLE
Bestbuy Bot Discussion

### DISCUSSION THREAD

**Discussion by TreborNamor** [TreborNamor/BestbuyBot#37] [General]
Welcome to the Bestbuy discussion board!

**Comment by bwbond87** [comment]
@TreborNamor in the `please_wait_enabled` block, should the call to `driver.refresh()` be there? I thought if you refreshed the page you would lose your place in the queue and have to start all over. If the "please wait" functionality on the website is all done via JS, can the refresh call just be removed?

edit: I think I may just be misunderstanding how `driver.refresh()` operates with selenium.

**Comment by TreborNamor** [comment]
I've heard of that before but I personally don't think refreshing the page
will lose your place in line. What is happening when you click the add to
cart button is an async  call to the server. The server remembers the call
and sets a promise to respond. If the item is available the add to cart
button will reappear when the server responds. I would leave the page
refresh function there. The bot is refreshing every 15 seconds to find a
certain class attribute that triggers the synchronous buying process. I've
successfully tested a purchase on the Ryzen 5800x with the current
settings. I highly recommend you to try it out. If you like you can always
set a higher timer on refresh for anywhere from 1-5 minutes. That's usually
how long it takes the add to cart button to reappear.

**Reply by bwbond87** [reply]
OK, thanks for clarifying. Would lowering that 15 second interval increase the chances of getting a GPU or is that roughly the interval at which the async call is triggered anyways?

**Comment by ToastyAvocado** [comment]
@TreborNamor I have followed your instructions on editing the code to match what I am looking for but when I click run then the file with the code a error pops up at the bottom saying Traceback (most recent call last):
  File "filename", line 1, in <module>
    import bs4
ModuleNotFoundError: No module named 'bs4'

Process finished with exit code 1

There is also 7 errors
No module named bs4
unresolved reference twilio
unresolved reference Client 
unresolved reference twilio
unresolved reference TwilioRestException 
unresolved reference webdriver_manager 
unresolved reference GeckoDriverManager

I am planning not to use twilio for message updates if that helps.

**Comment by TreborNamor** [comment]
You can change it to lower if you like. I just had it at 15 seconds as a
default. I wouldn't got lower than 3 seconds. It gives bestbuy time to
refresh web elements

**Reply by Kaltrops** [reply]
Hi Robert,

I wanted to thank you for a wonderful script, as well as to provide you with some data points.

Indeed your script worked marvelously well when trying to purchase a Ryzen cpu 2 days ago. However, despite its heroic efforts, the bot was not able to secure an RTX gpu yesterday. It managed to click twice several times, but every time it checked, the cart was empty.

Note that I'm not complaining! This is the closest I ever got to nabbing a card, so thank you so much for your work.

**Comment by TreborNamor** [comment]
Type, pip install bs4 in the terminal to fix error.

**Reply by ToastyAvocado** [reply]
I did this but the other 6 errors are still there and I can't run it

**Comment by Maxwell-Rosen** [comment]
Hi there,

I love the work and hope this helps me snag a graphics card for my new build. 
I've had an issue with the automatic log-in system within firefox. bestbuybot.py just has a line 
`print("Attempting to Login. Firefox should remember your login info to auto-login.")`
This is fine, but sometimes it will log me out after some time and asks for login after clicking checkout. It would be great if the code could get past this page if it appears by pushing the log-in button. I tried looking into that page, but it isn't as easy to find the buttons as on the product page. Not sure if this is an easy fix. I'm not experienced in how to use selenium, so maybe it's easier for you to look at. Another fix would be including some documentation for configuring Firefox to automatically log in or keep you logged in.

Also, I added a text notification when the add-to-cart button appears on BestBuy so I can monitor the progress of the code. This might help with knowing when to monitor Firefox in case it makes any mistakes. Maybe this is a change you might want to include in the next itteration. Through my testing, I found that sending a text doesn't delay the code in a significant way and provides better feedback. Also, I love the integration with Twillow. Great idea.

**Reply by Maxwell-Rosen** [reply]
Update: I fixed my first issue. It turns out, if you log in using google (which is what I used), then it automatically logs you in through this page. Another change I made is defining a function `text(str)` which uses the code which is repeated a few times to send texts through twillow. I also found a bug where after a while, the program won't be able to refresh and it leads to an error on the timer at the end of the while True loop for scanning all available cards and refreshing the page. Hope this helps!

**Comment by jscpark321** [comment]
Hi @TreborNamor 

Thanks for providing a detailed instruction on how to install the bestbuy bot. I followed each step and got it to work pretty smoothly. I just had a question regarding the URLs - does it only accept a single URL or is there a way to add multiple URLs of different products? I'm new to PyCharm, let alone coding, so any help would be much appreciated.

Thank you!

**Comment by VellichorPrism** [comment]
@TreborNamor Great job on the bot and thank you for making this available! The Best Buy bot seems to not work though, when in the queue, once the "Please Wait..." is replaced with the new "Add to Cart", the bot doesn't proceed and click the new "Add to Cart" and I have to do the rest manually. Do you know what could be the problem, or if there is a fix? Thank you!

**Comment by Kali123411** [comment]
Thanks for the bot. I just tested it and it appears to be working correctly. Will report back if I am able to purchase a GPU.

**Comment by johnleeforks** [comment]
hi, I'm getting an error with WebDriver Manager, could any body point me in the right direction?

.'(Default)').VersionInfo.ProductVersion was unexpected at this time.
Traceback (most recent call last):
  File "C:/Users/johnw/PycharmProjects/bestbuybot/bestbuy.py", line 243, in <module>
    driver = create_driver()
  File "C:/Users/johnw/PycharmProjects/bestbuybot/bestbuy.py", line 68, in create_driver
    web_driver = webdriver.Firefox(profile, options=options, executable_path=GeckoDriverManager().install())
  File "C:\Users\johnw\PycharmProjects\bestbuybot\venv\lib\site-packages\webdriver_manager\firefox.py", line 22, in __init__
    self.driver = GeckoDriver(version=version,
  File "C:\Users\johnw\PycharmProjects\bestbuybot\venv\lib\site-packages\webdriver_manager\driver.py", line 83, in __init__
    self.browser_version = firefox_version()
  File "C:\Users\johnw\PycharmProjects\bestbuybot\venv\lib\site-packages\webdriver_manager\utils.py", line 164, in firefox_version
    raise ValueError(f'Could not get version for Firefox with this command: {cmd}')
ValueError: Could not get version for Firefox with this command: (Get-Item (Get-ItemProperty 'HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths\firefox.exe').'(Default)').VersionInfo.ProductVersion

Process finished with exit code 1

**Reply by TheD0nald** [reply]
Following as I am getting this as well

**Reply by cdustevich1** [reply]
I fixed this by downloading the geckodriver.exe from [https://github.com/mozilla/geckodriver/releases](url) and putting in the directory 'C:\Users\gecko\'. The I changed the script as shown in this example: 
![BestBuyBotFix](https://user-images.githubusercontent.com/35354000/116626882-10bd0900-a91a-11eb-8588-875292bb0d8f.PNG). You can likely change the file path but this is just what I chose. I commented line 72 and copied and pasted (line 73) it then changed the executable path to "r'C:\path\to\file'". This location likely doesn't matter it's just for example. I am sure there is a better fix but this worked for me.

**Comment by Kali123411** [comment]
I had 4 instances of this script running and when I came home one of them had ceased with the following error:
![error](https://user-images.githubusercontent.com/64617127/116488639-11896880-a850-11eb-93f3-753059001ea9.JPG)

**Reply by TheD0nald** [reply]
Kali I noticed you posting in the newegg bot is it still functioning currently?

**Comment by TheD0nald** [comment]
https://pypi.org/project/webdriver-manager/#description

^^ see link above, looks like new version of webdriver dropped today and it shows the TESTING as Failing.

**Reply by Kali123411** [reply]
Yeah i'm getting errors related to webdriver

**Reply by TheD0nald** [reply]
im gonna roll the driver back to the previous version and see what I get

**Comment by TheD0nald** [comment]
Rolling back the driver to the previous version did work, however the bot is messing up when choosing to ship the item

**Comment by x5060** [comment]
So I was able to reinstall webdriver-manager to the previous version to get it to work. I have fought through a few other issues, but now I am running into an issue where the best buy site is using a non-clickable spinner after you click the button and selenium is erroring out on that.

```
  File "C:\Users\*\PycharmProjects\BestBuy-AutoBuy\venv\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 321, in execute    self.error_handler.check_response(response)

  File "C:\Users\*\PycharmProjects\BestBuy-AutoBuy\venv\lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 242, in check_response    raise exception_class(message, screen, stacktrace)

selenium.common.exceptions.ElementClickInterceptedException: Message: Element <button class="btn btn-lg btn-block btn-primary" type="button"> is not clickable at point (1478,370) because another element <div class="page-spinner page-spinner--out"> obscures it
```

**Reply by x5060** [reply]
Ok I was able to get  past this. I just put a time.sleep(2) just under the "try:" at line 83.

Did some more testing and we'll see what happens. Going for a reasonably priced 3080! Just need one for my personal build......

Also I did a price box on the search from $500-$750

**Reply by TheD0nald** [reply]
I have it running on a 3080 right now and not getting that error, obviously its just refreshing until something comes available. I have 90% of this built in powershell using selenium but I havent seen the best buy que system and need to test getting through that

Does anyone know if newegg requires any weird things to check out like best buy does?

**Comment by RCJeans** [comment]


**Reply by RCJeans** [reply]
i am very new to coding and tried to do some outside research for a solution and people said to just put an r before the quations where i entered the root location but that didnt work

**Reply by x5060** [reply]
I assume you're using the newer aggressive scripts. 
Your Line 41 should look something like this.

profile = webdriver.FirefoxProfile(r'C:\Users\Nick\AppData\Roaming\Mozilla\Firefox\Profiles\88t9nvrt.default-release')

**Comment by Detrivance** [comment]
It seems like there is a new [fulfillment](https://i.imgur.com/mLKmJRV.png) screen that was added to the BestBuy website. It now requires an extra step of clicking continue to payment where it enters in the CVV and the billing address as this is not stored on best buy's account information, only the shipping address is. The bot gets to this screen "enters in the CVV" and then stops.

Is there an update that will be happening for this change from BestBuy?

**Reply by CloudJanitor** [reply]
Put this below Shipping options as its own code block. Should correct that issue. Something that you will want to notice, I have had bad luck having it save Billing Information, so you will need another block of code to address that too. 

`               
 # Continue to Payment Information
                try:
                    print("\nProcessing Payment Information.\n")
                    wait2.until(EC.presence_of_element_located((By.XPATH, "//*[@class='btn btn-lg btn-block btn-secondary']")))
                    time.sleep(1)
                    driver_click(driver, 'xpath', 'btn btn-lg btn-block btn-secondary')

                except (NoSuchElementException, TimeoutException):
                    pass`

**Comment by Kali123411** [comment]
This script is totally broken for me now...

**Comment by TreborNamor** [comment]
Don't worry guys I'll keep the script updated soon. I've just been in a 5
month coding bootcamp that ends tomorrow. I haven't had alot of time to
keep up with it. I'll make sure to fix it. I have a new amazon script that
I've been working on. Will release it after testing for a few days.

**Reply by TheD0nald** [reply]
Trebor how do you test out the best buy script with the new que system if everything is sold out? I built this is powershell using selenium but im not sure how to buildout the que system part because everything is sold out and I havent physically experienced the que system

**Reply by CloudJanitor** [reply]
Do it against the 5800x, it is usually in stock, and it has the queue system in place.

https://www.bestbuy.com/site/amd-ryzen-7-5800x-4th-gen-8-core-16-threads-unlocked-desktop-processor-without-cooler/6439000.p?skuId=6439000

**Reply by TheD0nald** [reply]
Your goated bro

I have the full checkout process built and working with update billing information, I just needed to see the que process

**Reply by TheD0nald** [reply]
Hmm It hasnt became available for me all day is there another item to test maybe?

**Comment by Moto828** [comment]
> I had 4 instances of this script running and when I came home one of them had ceased with the following error:
> ![error](https://user-images.githubusercontent.com/64617127/116488639-11896880-a850-11eb-93f3-753059001ea9.JPG)

I am having the same issue... Backdating the driver did not fix this issue. Have you figured out a fix?

**Reply by Moto828** [reply]
Anyone?

**Reply by RadicalVictor** [reply]
Soft ban I'm guessing? If so found any way around it?

**Comment by TreborNamor** [comment]
You can experience the queue system by using the ryzen 5800x chip URL. It
still uses the queue system even though the item is in stock. I made the
script doing that.

**Reply by TheD0nald** [reply]
Also I have it clicking the initial add to cart button, then sleeping for 1 second and trying the click again, Eventually it seems after it doesnt become available for a period of time it kicks you out of que and you have to reclick add to cart to go back in que

**Comment by TheD0nald** [comment]
I have this working in powershell now

One note if the person has a best buy credit card saved and are using that for checkout it changes the final page for buying

**Reply by CloudJanitor** [reply]
Yea, i notice that, does not require the CVV to be verified. I ended up removing the bestbuy cc. Are you also experiencing bestbuy failing to saving billing address for future use? I had to make a programming block to correct this.

**Reply by TheD0nald** [reply]
yea for mine I just ask the users input for $streetAddress, $city, $state and $zip. Then it passes those variables to the website

**Reply by CloudJanitor** [reply]
Ha that is what i did. Its not pretty, but gets the job done.

`
# 4. Billing Address since BB is being a pos
fName = 'xxx'
lName = 'xxxxxx'
street = 'xxxxxxx'
city = 'xxxxxxx'
state = 'xxxx'
zipcode = 'xxxxxxx'

                # Trying Billing Address
                try:
                    print("\nTrying Billing Address Information.\n")
                    wait2.until(EC.presence_of_element_located((By.ID, "payment.billingAddress.firstName")))
                    firstNamefield = driver.find_element_by_id("payment.billingAddress.firstName")
                    firstNamefield.send_keys(fName)
                    wait2.until(EC.presence_of_element_located((By.ID, "payment.billingAddress.lastName")))
                    lastNamefield = driver.find_element_by_id("payment.billingAddress.lastName")
                    lastNamefield.send_keys(lName)
                    wait2.until(EC.presence_of_element_located((By.ID, "payment.billingAddress.street")))
                    streetfield = driver.find_element_by_id("payment.billingAddress.street")
                    streetfield.send_keys(street)                    
                    wait2.until(EC.presence_of_element_located((By.ID, "payment.billingAddress.city")))                    
                    cityfield = driver.find_element_by_id("payment.billingAddress.city")
                    cityfield.send_keys(city)
                    wait2.until(EC.presence_of_element_located((By.ID, "payment.billingAddress.zipcode")))                    
                    zipcodefield = driver.find_element_by_id("payment.billingAddress.zipcode")
                    zipcodefield.send_keys(zipcode)
                    wait2.until(EC.presence_of_element_located((By.ID, "payment.billingAddress.state")))
                    statefield = driver.find_element_by_id("payment.billingAddress.state")
                    statefield.send_keys(state)
                except (NoSuchElementException, TimeoutException):
                    pass

`

**Reply by TheD0nald** [reply]
im going to through both checkout into a function and ask the user to chose if they are using a best buy credit card or non best buy card

**Reply by TheD0nald** [reply]
Cloud do you know if the sold out button will turn yellow WITHOUT having to be constantly refreshed? Currently im refreshing every second and then trying to click the add to cart button

**Reply by CloudJanitor** [reply]
That I do not know :). Gotta be careful, BestBuy will cookie ban you if your refreshes are less than 20 seconds typically. (Full page refresh).

**Reply by TheD0nald** [reply]
Is that what you do? Refresh every 20 seconds?

**Reply by CloudJanitor** [reply]
I am using the default that Trebor has, which is like 15+5?

**Reply by TheD0nald** [reply]
do you think 20 seconds will be fast enough to get into the que?

**Reply by CloudJanitor** [reply]
I am not sure you have a choice, you can experiment with lower timers, but in my experience, it has always resulted in a cookie ban.

**Reply by TheD0nald** [reply]
ive been runnin 1 second refreshes for 2.5 days and havent got a ban lol.

Also, have you had success getting a card with this?

**Reply by TheD0nald** [reply]
Cloud do you have discord, id love to chat with you I think we could help each other

**Reply by CloudJanitor** [reply]
Sure, hit me up @ thePoison#9636

**Reply by TheD0nald** [reply]
Sent

**Reply by Maxwell-Rosen** [reply]
Would it be possible to update this with a field for inputting your apartment number?

**Reply by brendanteske** [reply]
@TheD0nald can you expand on this cookie ban bit? I have been having SLOW AF load times on BB now... Is that what happened to me?? Is there a solution?

**Reply by Maxwell-Rosen** [reply]
I haven't experienced a cookie ban and I've been running this code for 8+ hours/day on-and-off for about 3 weeks. After it runs for about 4 hours, it gets a bit slower sometimes. I have run it on my shitty laptop and also my beefy desktop. The shitty laptop runs into an issue where it can't refresh the page as fast. My desktop handles it just fine though. My recommendation is just to stop the code every few hours to give your computer a break.

**Comment by Kaltrops** [comment]
Nope. There are too many ways for things to go wrong, and they have nothing
to do with the script. For example, I managed to get 1 GPU in the cart, but
then the GPU was not available at my store (or any other store for that
matter). Nothing the script can do about such nonsense.

On Thu, May 13, 2021 at 9:36 PM Moto828 ***@***.***> wrote:

> Anyone?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/TreborNamor/Agressive-Store-Bots/discussions/37#discussioncomment-736831>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT4ULFJONUQDWWK2NEOHVV3TNSSGVANCNFSM43GH3MOQ>
> .
>


-- 
Simon

**Comment by gkollaba** [comment]
So i have been trying to execute the script. It opens the firefox and the item page. I was testing with an item available and as checking if it actually adds the item to the cart and checks out. Even though the add to cart pops, it keeps refreshing the page. Anyone else had this issue? What do i do about this?

**Reply by PyrosGaming** [reply]
I can assume on amazon? if so The add to cart button is iffy and you have to mess with the class button when you inspect the cart button, maybe add an extra step in there so it thinks your legit, I had a similar issue with newegg and no one would respond so I figured it out for my self. once you watch a few videos its trail in error at that point. i already bought some books for python because this guy whos admin of this page has inspired me to get more into it, Its kinda fun when you start to realize what everything dose. The error with Newegg is you have to add in a extra step because the website acts differently when using a script for it.

**Reply by gkollaba** [reply]
It is acutally bestbuy. I'll debug it tonight and see how its working. But everyone here said they followed his instructions and it worked perfectly fine so i am wondering if i was doing something wrong.

**Comment by abigdreamer** [comment]
I'm getting 2 items in my cart. Testing with this item. url = 'https://www.bestbuy.com/site/sandisk-ultra-32gb-usb-3-0-flash-drive-black/9211069.p?skuId=9211069' -- Always 2 items. === Sometimes it does not checkout. 1 time it did. weird.

**Reply by Seanmatthews** [reply]
Yeah this script is extremely fragile. I don't even know how it would work, given that it never presses the "Checkout" button.

**Comment by PrinceTvlip** [comment]
I have the script working fine, but after about an hour it does crash. What would the reason be for this?

**Reply by PrinceTvlip** [reply]
the error code
selenium.common.exceptions.TimeoutException: Message: TimedPromise timed out after 300000 ms

**Reply by Maxwell-Rosen** [reply]
Where the error occurred, I think, is in refreshing the page before it was done loading. I've fixed this with this code 

`      try:
          time_sleep(5, driver)
      except:
          print('Error refreshing the page. Trying again')
          searching_for_card(driver)
`

Put this to replace the `time_sleep(5,driver)` at the end of the `searching_for_card()` function. In my version, which I have modified, it is line 316

**Reply by PrinceTvlip** [reply]
I'm sorry, I don't know what you mean by replacing time_sleep(5,driver) at the END of searching_for_card(). The line I have time_sleep(5, driver) is only line 241. This is what the end of my code looks like

time_sleep(5, driver) 


if __name__ == '__main__':
    driver = create_driver()
    searching_for_card(driver)

**Reply by Moto828** [reply]
Yes I believe that's the one he is referring to. It's line 239 on the original code.

**Reply by Maxwell-Rosen** [reply]
Yes. It's that time_sleep() function. Sorry for being unclear

**Reply by PrinceTvlip** [reply]
It's okay, I'm very new at this so a lot of it is just me wrapping my head around stuff. I did try and replace the time_sleep() function with yours;
try: time_sleep(5, driver) except: print('Error refreshing the page. Trying again') searching_for_card(driver)
but when replacing this function, it breaks the code and will not run. I assume it's a matter of me putting the code in the right line.

**Reply by Maxwell-Rosen** [reply]
Nah. Python is really sensitive to returns. I think posting the code changed the new lines. It should be 

Try:
   Time sleep
Except:
   Print
   Searching for card

But with the proper code. I can send the proper indented version when I'm home if that doesn't make sense

**Reply by Moto828** [reply]
Thanks mrquell for the solution! Changed the load timeout to 1ms to replicate the timeout error and after adding the code the timeout error is gone.

**Reply by PrinceTvlip** [reply]
I just got it to work as well. I'll update if it does crash within 3 hours! And just wondering, what refresh times do you guys use? It comes at 5 but I've been using 3. Can I go lower?

**Reply by Maxwell-Rosen** [reply]
I tried using 3 and 1 second, but it produced more errors since it gave the page less time to load. I left it at 5 and it didn't give too many errors. I guess with this fix it will fix that, but still it's not good to error. Also, 5 seconds is pretty fast compared to other stock checkers I've seen

**Reply by Moto828** [reply]
Luckily I have been able to run at 5s with only 1 timeout per day but ideally would prefer to run continuously without stopping even with errors it's better than completely stopping.

**Reply by Maxwell-Rosen** [reply]
This will fix that time-out issue. It rarely errors for me, but it does occasionally. It's noted in the log every time it errors. Not every line in the main section for refreshing the page has `try` statements, and that would help fix the errors from refreshing constantly

**Reply by Moto828** [reply]
I would avoid reducing under 5s since it will likely get your ip flagged for a ban. I have tested running a vpn and they usually soft ban the ip after some time... Even their developer api is limited to 5s requests before getting locked out.

**Reply by PrinceTvlip** [reply]
After a day of testing, it has still crashed on me multiple times. Sometimes the countdown from 5 to 0 just stops at 0 and will not error and/or continue to refresh the page.

This is two of the errors I did get:
RecursionError: maximum recursion depth exceeded while calling a Python object
chromeWindow is null

Edit: I use firefox, not chrome.

**Reply by Maxwell-Rosen** [reply]
You might not have the code indented properly because it looks like it's triggering the extra searching_for_card() function too many times. I just want to confirm this. Now that I'm back at my computer, it's easier for my to screenshot and send exactly the code.

![image](https://user-images.githubusercontent.com/48170600/119665072-ffc0c400-be01-11eb-9455-91d3ff7dca9c.png)

**Reply by PrinceTvlip** [reply]
ur a life saver

**Reply by abigdreamer** [reply]
Something that helps a lot is to compile into 1 file. 
pyinstaller  filename.py --onefile

Do this from the command line in terminal inside the bot folder. 
Then run the file with ./filename   

Way way faster than from pycharm. 
Bots would freeze before I did this. 
Somebody should add this to the readme for all bots.

**Reply by Moto828** [reply]
Are you running the script headless?

**Comment by ceddy-cedd** [comment]
Heres my stack trace:
Traceback (most recent call last):
File "/Users/cedric/Desktop/bestbuybot/botter.py", line 257, in
searching_for_card(driver)
File "/Users/cedric/Desktop/bestbuybot/botter.py", line 252, in searching_for_card
time_sleep(5, driver)
File "/Users/cedric/Desktop/bestbuybot/botter.py", line 94, in time_sleep
driver.execute_script('window.localStorage.clear();')
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/selenium/webdriver/remote/webdriver.py", line 634, in execute_script
return self.execute(command, {
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/selenium/webdriver/remote/webdriver.py", line 319, in execute
response = self.command_executor.execute(driver_command, params)
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/selenium/webdriver/remote/remote_connection.py", line 374, in execute
return self._request(command_info[0], url, body=data)
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/selenium/webdriver/remote/remote_connection.py", line 397, in _request
resp = self._conn.request(method, url, body=body, headers=headers)
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/request.py", line 78, in request
return self.request_encode_body(
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/request.py", line 170, in request_encode_body
return self.urlopen(method, url, **extra_kw)
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/poolmanager.py", line 375, in urlopen
response = conn.urlopen(method, u.request_uri, **kw)
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/connectionpool.py", line 783, in urlopen
return self.urlopen(
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/connectionpool.py", line 783, in urlopen
return self.urlopen(
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/connectionpool.py", line 783, in urlopen
return self.urlopen(
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/connectionpool.py", line 755, in urlopen
retries = retries.increment(
File "/Users/cedric/Library/Python/3.8/lib/python/site-packages/urllib3/util/retry.py", line 574, in increment
raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='127.0.0.1', port=51617): Max retries exceeded with url: /session/29a83df0-3b88-8740-a61b-613a4daff456/execute/sync (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x10da23820>: Failed to establish a new connection: [Errno 61] Connection refused'))

Any idea on how to fix this?

**Reply by ceddy-cedd** [reply]
Gets all the way to checkout but then crashes just after cvv instance and doesnt place order for example items

**Comment by raymondcg** [comment]
It looks like the bot if it hits best buy too much will get your ip a soft ban. 

Meaning if you try to add to cart it gives a generic error message. I noticed using my phone with the same IP, as soon as i flipped from wifi to mobile and got a new IP address the generic error went away.

I am not sure if its only the queued products it will do the error on, how to detect a soft ban besides the error, or what is the ban trigger level.

**Reply by Moto828** [reply]
For me easiest way to check if soft ban is present is trying to add a normal item to cart (I use an HDMI product) if you get the error than soft ban is active.

**Comment by Seanmatthews** [comment]
The script does not ever press the "Checkout" button before it tries to set shipping.

**Comment by shenjianan97** [comment]
@TreborNamor Thanks for the bot! I just used the 5800x page for testing. The code doesn't work if the item is added when the add to cart button is clicked for the first time. Maybe shouldcheck it in the code,

**Comment by JashQonPC** [comment]
I noticed each time firefox refreshes, it builds up memory usage on ram and I have to close out the browser and restart the script for the ram usage to go back down, but as it refreshes the ram builds back up again.  Once I hit full ram capacity, the script stops. Is there anyway to close out the browser and restart the page instead of just refreshing it? Or could it be something wrong with my firefox?

**Reply by Moto828** [reply]
I am not having this issue. What IDE software and version are you using?

**Reply by JashQonPC** [reply]
Im using Pycharm Community Edition 2021.2

**Reply by Moto828** [reply]
Hope this helps you solve your issue. I am using Pycharm 2021.1.1 / Firefox 89.0.2. Additionally readme.md file was updated few months back indicating to use pip install webdriver-manager==3.4.0.

**Reply by JashQonPC** [reply]
Yup, it was my firefox version that was the problem. Had to down grade to 89.0.2. My memory build up has now been fixed. Thanks a bunch.

**Comment by Moto828** [comment]
Anyone know if there's another product to test the code? 5800x is sold out and cannot find another product using the queue system.

**Comment by arcticowlnjm** [comment]
hey I am having issues with the bot adding more than one of the item to my cart. any fix for that?

**Reply by Moto828** [reply]
What are you trying to add? note that some items are limited to 1 per person especially for items in queue.

**Comment by JamesRakovsky** [comment]
Hey the code randomly breaks sometimes, anyone have a fix

**Reply by tagerjump** [reply]
Around line 137, update add_to_cart_button to 
```
add_to_cart_button = soup.find('button', {
                'class': 'c-button c-button-primary c-button-lg c-button-block c-button-icon c-button-icon-leading add-to-cart-button'})
```

**Reply by tagerjump** [reply]
You'll need also update 'credit-card-cvv' to 'cvv' around line 225
Also you should know that this script will buy two count of the item, and checkout all goods in your cart. So remember to clear you cart before run it in non-testing mode.

**Comment by DrWolvezs** [comment]
> Welcome to the Bestbuy discussion board!

Can I get banned for it refreshing so much

**Comment by DrWolvezs** [comment]
The best buy bot won't click add to cart when it's in stick is there a fix

**Comment by RasTooten** [comment]
when running the code it gets to  Add To Cart Button Found! but then does nothing further. I also have 1 warning (Unused import statement 'WebDriverException')

**Reply by RasTooten** [reply]
I have tried installing multiple versions of WebDriverException with no success.

**Reply by crazywowmen** [reply]
pls see my answer below

**Comment by Maxwell-Rosen** [comment]
Hello. I'm rerunning this, and it looks like the add to cart button has changed. When I run the script, it just keeps refreshing the page. I'm not super confident in editing the code, but It looks to be an issue with this line.
`add_to_cart_button = soup.find('button', {
               'class': 'btn btn-primary btn-lg btn-block btn-leading-ficon add-to-cart-button'})`
I think it has to do with the class being none for the actual button now. Not sure. Could someone suggest a fix?

**Reply by crazywowmen** [reply]
as of rn, you wanna update to this
```
add_to_cart_button = soup.find('button', {
                'class': 'c-button c-button-primary c-button-lg c-button-block c-button-icon c-button-icon-leading add-to-cart-button'})
```

but bb has detected this's a bot, the whole program no longer works for me. i can't log in to my account with the pop out firefox window from running the bot

**Comment by milesk182** [comment]
Hello, im new to all this and prepping to beat scalpers to get a 50 series card if i can. Im getting the following after getting bs4 installed and the webdriver installed:

I get the following error message.

Traceback (most recent call last):
  File "C:\PyCharm\pythonProject\rtx5090bot\rtx5090bot\spiders\rtx5090bot.py", line 257, in <module>
    driver = create_driver()
             ^^^^^^^^^^^^^^^
  File "C:\PyCharm\pythonProject\rtx5090bot\rtx5090bot\spiders\rtx5090bot.py", line 62, in create_driver
    web_driver = webdriver.Firefox(profile, options=options, executable_path=GeckoDriverManager().install())
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: WebDriver.__init__() got multiple values for argument 'options'

line 257  driver = create_driver()
line 62    web_driver = webdriver.Firefox(profile, options=options, executable_path=GeckoDriverManager().install())

Any help would be greatly appreciated

**Comment by SkitVibes** [comment]
Hey does this still work for the 5090 release tomorrow?

**Comment by milesk182** [comment]
I haven't been able to get it still at all. I think first drop is Thursday
at 7am.

**Comment by Skelotone** [comment]
**Current Bug**
It was kind of working the last drop, the only issue I saw was that it would get stuck on the sold out screen. Now it does not work at all, When you go to the item you want (Founders 5090 in this case), when it refreshes it just takes you to the reviews page then no longer refreshes.
